pydantic_settings/main.py

```diff
@@ -134,98 +153,166 @@ def parse_env_var(cls, field_name: str, raw_val: str) -> Any:
     __config__: ClassVar[Type[Config]]
 
 
-class InitSettingsSource:
+class PydanticBaseSource(ABC):
+    def __init__(self, settings_cls: Type[BaseSettings]):
+        self.settings_cls = settings_cls
+        self.config = settings_cls.__config__
+        self.case_sensitive = self.config.case_sensitive
+
+    @abstractmethod
+    def get_field_value(self, field: ModelField) -> Any:
+        pass
+
+    def prepare_field(self, field_name: str, field: ModelField, value: Any) -> Any:
+        if field.is_complex():
+            return json.loads(value)
+        return value
+
+    def __call__(self) -> Dict[str, Any]:
+        d: Dict[str, Any] = {}
+
+        for name, field in self.settings_cls.__fields__.items():
+            field_value = self.get_field_value(field)
+
+            try:
+                field_value = self.prepare_field(name, field, field_value)
+            except ValueError as e:
+                raise SettingsError(f'error parsing value for field "{name}"') from e
```

Comment:
this should probably say which source we're getting data from to to ease debugging.

Reply:
Setting source added to error message and tests updated